go/store/nbs/journal_writer.go

```diff
@@ -133,11 +146,50 @@ func createJournalWriter(ctx context.Context, path string) (wr *journalWriter, e
 		return nil, fmt.Errorf("expected file journalOffset 0, got %d", o)
 	}
 
-	return &journalWriter{
+	sendEg, _ := errgroup.WithContext(ctx)
+	recvEg, ctx := errgroup.WithContext(ctx)
+	indexChan := make(chan any, indexChanSize)
+	wr = &journalWriter{
 		buf:     make([]byte, 0, journalWriterBuffSize),
 		journal: f,
 		path:    path,
-	}, nil
+		recvEg:  recvEg,
+		sendEg:  sendEg,
+		indexCh: indexChan,
+	}
+	recvEg.Go(func() error {
+		return wr.recvIndexRecords(ctx, indexChan)
+	})
+
+	return wr, nil
+}
+
+func (wr *journalWriter) recvIndexRecords(ctx context.Context, c chan any) error {
+	for {
+		select {
+		case <-ctx.Done():
+			return context.Cause(ctx)
+		case obj, ok := <-c:
+			if !ok {
+				return nil
+			}
+			var err error
+			switch l := obj.(type) {
+			case lookup:
+				wr.batchCrc = crc32.Update(wr.batchCrc, crcTable, l.a[:])
+				err = writeIndexLookup(wr.indexWriter, l)
+			case lookupMeta:
+				l.checkSum = wr.batchCrc
+				wr.batchCrc = 0
+				err = writeJournalIndexMeta(wr.indexWriter, l.latestHash, l.batchStart, l.batchEnd, l.checkSum)
+			default:
+				err = fmt.Errorf("invalid index channel object")
+			}
+			if err != nil {
+				return err
+			}
```

Comment:
There seems to a problem with the error handling and communication here. Before this change, an error writing the index would cause the write call to report an error. Now, this write call and future write calls will not report an error, and we will not see an error at Close() as currently written either. Which is, maybe fine, but we will also stop reading from `indexChan`, and that does not seem fine, since the write path expects to be able to write to `indexChan`.

You've already taken the write lock on the sender side. An easy way forward is to just set `indexChan` to nil and report the error in a shared variable `indexWriteErr` and make future writes return that thing? Potentially restart the goroutine on a future write if you want to get back the old semantics?